src/main/java/io/supertokens/storage/mysql/queries/UserRoleQueries.java

```diff
@@ -0,0 +1,222 @@
+import java.sql.SQLException;
+import java.util.ArrayList;
+
+import static io.supertokens.storage.mysql.QueryExecutorTemplate.execute;
+import static io.supertokens.storage.mysql.QueryExecutorTemplate.update;
+
+public class UserRoleQueries {
+
+    public static String getQueryToCreateRolesTable(Start start) {
+        String tableName = Config.getConfig(start).getRolesTable();
+        // @formatter:off
+            return "CREATE TABLE IF NOT EXISTS " + tableName + " ( role VARCHAR(255) NOT NULL, " +
+                    "PRIMARY KEY(role))";
+        // @formatter:on
+    }
+
+    public static String getQueryToCreateRolePermissionsTable(Start start) {
+        String tableName = Config.getConfig(start).getUserRolesPermissionsTable();
+        // @formatter:off
+        return "CREATE TABLE IF NOT EXISTS " + tableName + " ( " +
+                "role VARCHAR(255) NOT NULL, " +
+                "permission VARCHAR(255) NOT NULL, " +
+                "PRIMARY KEY (role, permission), " +
+                "FOREIGN KEY (role) REFERENCES " +
+                Config.getConfig(start).getRolesTable()+ "(role) ON DELETE CASCADE )";
+        // @formatter:on
+    }
+
+    public static String getQueryToCreateRolePermissionsPermissionIndex(Start start) {
+        return "CREATE INDEX role_permissions_permission_index ON "
+                + Config.getConfig(start).getUserRolesPermissionsTable() + "(permission);";
+    }
+
+    public static String getQueryToCreateUserRolesTable(Start start) {
+        String tableName = Config.getConfig(start).getUserRolesTable();
+        return "CREATE TABLE IF NOT EXISTS " + tableName + "( user_id VARCHAR(128) NOT NULL, "
+                + "role VARCHAR(255) NOT NULL," + "PRIMARY KEY (user_id, role)," + "FOREIGN KEY (role) REFERENCES "
+                + Config.getConfig(start).getRolesTable() + "(role) ON DELETE CASCADE)";
+    }
+
+    public static String getQueryToCreateUserRolesRoleIndex(Start start) {
+        return "CREATE INDEX user_roles_role_index ON " + Config.getConfig(start).getUserRolesTable() + "(role)";
+    }
+
+    public static void addRoleToUser(Start start, String userId, String role)
+            throws SQLException, StorageQueryException {
+        String QUERY = "INSERT INTO " + Config.getConfig(start).getUserRolesTable() + "(user_id, role) VALUES(?, ?)";
+        update(start, QUERY, pst -> {
+            pst.setString(1, userId);
+            pst.setString(2, role);
+        });
+    }
+
+    public static boolean createNewRoleOrDoNothingIfExists_Transaction(Start start, Connection con, String role)
+            throws SQLException, StorageQueryException {
+        // We use IGNORE here since we want to ignore any conflict(duplicate key exception) that might be thrown.
+        // We do not use ON DUPLICATE UPDATE since it will return false positives when a role already exists
+        String QUERY = "INSERT IGNORE INTO " + Config.getConfig(start).getRolesTable() + " VALUES(?)";
```

Comment:
I don't think using `INSERT IGNORE` is good. Cause it allows you to insert null values into a non null column (based on https://stackoverflow.com/a/548570/3867175 answer)

Instead, you should use `... ON DUPLICATE KEY UPDATE` method. According to this answer (https://stackoverflow.com/a/7853634/3867175) it still shows number of rows affected as 0.

What do you mean by "return false positives"?